policy: track policy rule origin per selector

bimmlerd opened the pull request:
The current L4Filter.DerivedFromRules track a list of policy rules (identified through their labels). As the policy subsystem is (roughly) organised L4->L3->L7, we coalesce all rules which match the same flows on the L4 layer into a single L4Filter struct. Hence, when trying to understand which of the coalesced rules is the reason for a flow verdict, we don't know which of said list of rules truly matches a flow. (We never miss a rule which matches, but we list too many of them.)

This patch changes the list to be a map of lists, keyed by cached selector. In effect, this adds L3 information to the rule origin tracking. No attempt is made (yet) to also understand L7 semantics, due to the complexity involved in dealing with the proxy.

bimmlerd committed:
policy: track policy rule origin per selector

The current L4Filter.DerivedFromRules track a list of policy
rules (identified through a LabelArray). As the policy subsystem is
(roughly) organised L4->L3->L7, we coalesce all rules which match the
same flows on the L4 layer. Hence, when trying to understand which rule
is the origin for a flow verdict, we don't know which of the list of
rules truly matches a flow.

This patch changes the list to be a map of lists, keyed by cached
selector. In effect, this adds L3 information to the rule origin
tracking. No attempt is made to also understand L7 semantics, as those
are handed off to the proxy. Still, this improves the accuracy of the
verdict to rule matching, as the endpoints involved are now considered
as well.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

gandro commented on pkg/policy/l4.go:
```diff
@@ -1229,17 +1243,25 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
 	for _, v := range l4.Ingress {
+		derivedFrom := labels.LabelArrayList{}
+		for _, rules := range v.RuleOrigin {
+			derivedFrom = append(derivedFrom, rules...)
```
Does the fact that we iterate over a map here change the order of labels in `derivedFrom`? Or was it already not a stable order before?

bimmlerd replied:
good catch: It was a stable oder before; sorted even.

gandro commented on pkg/policy/l4.go:
```diff
@@ -1229,17 +1243,25 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
 		ingress = append(ingress, &models.PolicyRule{
 			Rule:             v.Marshal(),
-			DerivedFromRules: v.DerivedFromRules.GetModel(),
+			DerivedFromRules: derivedFrom.GetModel(),
```
Would it make sense to extend the Cilium API to also expose the RuleOrigin map here? That would give us a bit more introspection in sysdumps etc.

bimmlerd replied:
I've looked into this a bit: OpenAPI allows only string keys for maps. As a result, we'd have to encode the `CachedSelectors` (acting as keys in `RuleOrigin`) as strings. There's code to do this, but relying on it here would effectively add that to our public API surface, which seems suboptimal. Thoughts?

jrajahalme replied:
Make sure we expose whatever is the canonical string representation for an k8s EndpointSelector? Not necessarily the same we use internally.

bimmlerd replied:
Traced the implementation of CachedSelector.String() back to k8s, so I think this is fine as implemented.

gandro reviewed (approved):
Awesome work! I'm not super familiar with the policy engine, but the changes look correct to me.

Left one comment regarding the public API.

tommyp1ckles reviewed (approved):
k8s changes lgtm

aditighag reviewed (commented):
IIUC, this patch set will enable users to get information about matched L3 policy rules in addition to L4 ones. This is super handy! How's this exposed to end users? Can you attach a snippet to the PR/commit description to show the enhancement?

Edit: Synced up offline, and we can decouple these changes from how they are consumed for now.  I echo Sebastian's suggestion around having some way of inspecting the state - https://github.com/cilium/cilium/pull/23811#discussion_r1108335821.

aditighag reviewed (commented):
The logic seems sane to me at a high level, but it'll be great if we could add unit test cases to get some more confidence about the changes (and to catch any regressions in future). 

I am also requesting a review from @jrajahalme as he is intimately familiar with some of the cached selector logic. I see your note around not addressing L7 policies in this PR, but I wonder if the visibility policies are created with an independent cached selector or not.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -961,16 +972,19 @@ func addL4Filter(policyCtx PolicyContext,
+	for cs, newLabels := range filterToMerge.RuleOrigin {
+		if existingLabels, ok := existingFilter.RuleOrigin[cs]; ok {
+			if !existingLabels.DeepEqual(&newLabels) {
+				existingFilter.RuleOrigin[cs] = append(existingFilter.RuleOrigin[cs], newLabels...)
```
This seems to append the whole `LabelArrayList` for the `cs` from `filterToMerge` to `existingFilter`, possibly leading `existingFilter.RuleOrigin[cs]` containing duplicates, especially if two rules have the same `ruleLabels`. Also looks like we always did this, but it would be nice to not have duplicate label arrays in the list?

jrajahalme replied:
Maybe add a `LabelArrayList.Merge()` function to `pkg/labels/arraylist.go` and use it from here?

bimmlerd replied:
implemented Merge as suggested
As is it doesn't try to deduplicate the existing lists, instead just avoids introducing new duplication. Deduplication makes the code a fair bit more complex, but can be done if we think it's worth it.

jrajahalme replied:
I think not introducing duplicates matches the prior implementation, deduplication can be left for later if we want to address it.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -1229,17 +1243,27 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
 	for _, v := range l4.Ingress {
+		derivedFrom := labels.LabelArrayList{}
+		for _, rules := range v.RuleOrigin {
+			derivedFrom = append(derivedFrom, rules...)
```
This could also lead to duplicates, when the same `ruleLabels` applies to multiple selectors in `v.RuleOrigin`.

jrajahalme commented on pkg/policy/l4.go:
```diff
@@ -1229,17 +1243,27 @@ func (l4 *L4Policy) GetModel() *models.L4Policy {
 	for _, v := range l4.Egress {
+		derivedFrom := labels.LabelArrayList{}
+		for _, rules := range v.RuleOrigin {
+			derivedFrom = append(derivedFrom, rules...)
```
Same deduping needed here.

bimmlerd replied:
done

jrajahalme reviewed (changes requested):
GetModel() change may lead to duplicates, as rule's labels are in the `RuleOrigin` for each cached selector in the rule. To keep the old API we'll need to not append the same label array twice.

bimmlerd committed:
labels: introduce LabelArrayList.Merge

It rarely makes sense for LabelArrayLists to contain duplicate
LabelArrays. Introduce Merge to merge new LabelArrays into an existing
LabelArrayList without introducing duplicates.

Note that _existing_ duplicates, in either LabelArrayList, will lead to
duplicates in the output. Deduplicating both lists on merge would
complicate the code.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

bimmlerd committed:
api: expose RuleOrigin for policy origin tracking

RuleOrigin maps L3 selectors to policy rule labels for L4 filters.
Exposing this in the API allows for better introspection.

Signed-off-by: David Bimmler <david.bimmler@isovalent.com>

nathanjsweet reviewed (approved):
LGTM for API changes.
